protostar/starknet/cheatable_cached_state.py

```diff
@@ -174,6 +177,32 @@ def unregister_expected_call(
             if not data_for_address:
                 del self.expected_contract_calls[contract_address]
 
+    def register_expected_call_cairo(
+        self, contract_address: Address, function_name: str, calldata: CairoData
+    ):
+        if self.expected_contract_calls_cairo.get(contract_address):
+            self.expected_contract_calls_cairo[contract_address].append(
+                (function_name, calldata)
+            )
+        else:
+            self.expected_contract_calls_cairo[contract_address] = [
+                (function_name, calldata)
+            ]
+
+    def unregister_expected_call_cairo(
+        self, contract_address: Address, function_name: str, calldata: CairoData
+    ):
+        data_for_address = self.expected_contract_calls_cairo.get(contract_address)
+        if not data_for_address:
+            return
+        for index, (stored_function_name, stored_calldata) in enumerate(
+            data_for_address
+        ):
+            if stored_function_name == function_name and stored_calldata == calldata:
+                del data_for_address[index]
```

Comment:
Well, ok. I came up with something like this:
```
self.expected_contract_calls_cairo[contract_address] = [
    (stored_function_name, stored_calldata)
    for (stored_function_name, stored_calldata) in self.expected_contract_calls_cairo[contract_address]
    if stored_function_name != function_name or stored_calldata != calldata
]
```
Instead of this
```
for index, (stored_function_name, stored_calldata) in enumerate(
    data_for_address
):
    if stored_function_name == function_name and stored_calldata == calldata:
        del data_for_address[index]
```

I don't know, for me, it's far worse to read but I don't mind that much, also it adds a memory overhead (additional alloc instead of unlink? As far as I understand) but we don't operate on big data sets so that's fine I guess.

I will follow your advice but honestly, I'm not convinced this is better.

Reply:
[address various feedback comments](https://github.com/software-mansion/protostar/pull/1363/commits/9b066b37a326344e44dbcc224db465ed14eddb4e)